django/contrib/messages/storage/cookie.py

```diff
@@ -172,3 +214,39 @@ def _decode(self, data):
         # with the data.
         self.used = True
         return None
+
+
+def bisect_keep_left(a, fn):
+    """
+    Finds the index of the first element from the start of the array
+    that verifies the given condition.
+    The function is applied from the start of the array to the pivot.
+    """
+    lo = 0
+    hi = len(a)
+
+    while lo < hi:
+        mid = (lo + hi) // 2
+        if fn(a[: mid + 1]):
+            hi = mid
+        else:
+            lo = mid + 1
+    return lo
+
+
+def bisect_keep_right(a, fn):
+    """
+    Finds the index of the first element from the end of the array
+    that verifies the given condition.
+    The function is applied from the pivot to the end of array.
+    """
+    lo = 0
+    hi = len(a)
+
+    while lo < hi:
+        mid = (lo + hi) // 2
+        if fn(a[mid:]):
+            lo = mid + 1
+        else:
+            hi = mid
+    return lo
```

Comment:
OK, there is a comment in PR description :thinking:

Reply:
Yes, I tried using the functions from the `bisect` module, but since we need to apply the function to a subset of the input array, either from the start or the end, it's quite specific.
But the implementations for the `bisect_keep_left` and `bisect_keep_right` are based off the [CPython implementation](https://github.com/python/cpython/blob/main/Lib/bisect.py#L44-L49).